Parser/asdl_c.py

```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
+                Py_DECREF(expecting);
+                return -1;
+            }
+        }
+    }
+    // check that the remaining fields or attributes would be filled
+    if (dict) {
+        Py_ssize_t pos = 0;
+        PyObject *key, *value;
+        while (PyDict_Next(dict, &pos, &key, &value)) {
+            // Mark fields or attributes that are found on the instance
+            // as non-mandatory. If they are not given in 'kwargs', they
+            // will be shallow-coied; otherwise, they would be replaced
+            // (not in this function).
+            if (PySet_Discard(expecting, key) < 0) {
+                Py_DECREF(expecting);
+                return -1;
+            }
+        }
+        if (attributes) {
+            // Some attributes may or may not be present at runtime.
+            // In particular, now that we checked whether 'kwargs'
+            // is correct or not, we allow any attribute to be missing.
+            //
+            // Note that fields must still be entirely determined when
+            // calling the constructor later.
+            PyObject *optionals = PySet_New(attributes);
+            if (optionals == NULL) {
+                Py_DECREF(expecting);
+                return -1;
+            }
+            Py_ssize_t pos = 0;
+            PyObject *item;
+            Py_hash_t hash;
+            while (_PySet_NextEntry(optionals, &pos, &item, &hash)) {
+                if(PySet_Discard(expecting, item) < 0) {
+                    Py_DECREF(expecting);
+                    Py_DECREF(optionals);
+                    return -1;
+                }
+            }
+            Py_DECREF(optionals);
+        }
+    }
+    // Now 'expecting' contains the fields or attributes
+    // that would not be filled inside ast_type_replace().
+    Py_ssize_t m = PySet_GET_SIZE(expecting);
+    if (m > 0) {
+        PyObject *names = PyList_New(m);
+        if (names == NULL) {
+            Py_DECREF(expecting);
+            return -1;
+        }
+        Py_ssize_t i = 0, pos = 0;
+        PyObject *item;
+        Py_hash_t hash;
+        while (_PySet_NextEntry(expecting, &pos, &item, &hash)) {
+            PyObject *name = PyObject_Repr(item);
+            if (name == NULL) {
+                Py_DECREF(expecting);
+                Py_DECREF(names);
+                return -1;
+            }
+            // steal the reference 'name'
+            PyList_SET_ITEM(names, i++, name);
+        }
+        Py_DECREF(expecting);
+        if (PyList_Sort(names) < 0) {
+            Py_DECREF(names);
+            return -1;
+        }
+        PyObject *sep = PyUnicode_FromString(", ");
+        if (sep == NULL) {
+            Py_DECREF(names);
+            return -1;
+        }
+        PyObject *str_names = PyUnicode_Join(sep, names);
+        Py_DECREF(sep);
+        Py_DECREF(names);
+        if (str_names == NULL) {
+            return -1;
+        }
```

Comment:
Same question here, but do we have some utility function for the Python equivalent:

```py
names = ', '.join(map(repr, sorted(iterable)))
```

If not, would it make sense to provide such functions (only to be used inside CPython)?

Reply:
I think what you have is fine. An alternative approach could be to use PyUnicodeWriter but I don't know if you'll end up with less code.

Reply:
I'll keep this one then. I still need to construct that sorted list (maybe you have a shortcut for `sorted(map(repr, iterable))`? if not, then I'll use the current code.

Reply:
I'd probably do PySequence_List, then PyList_Sort, then iterate over the list and add to the writer.

Reply:
(But again, not convinced that's going to be simpler than what you have, and it does add an extra intermediate list.)

Reply:
Yes. Maybe I'll suggest some helper function but I'll stick with what I did here. I prefer not to allocate too many temporary objects...